extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,423 @@
+   * @return {boolean}
+   * @private
+   */
+  onRejection_(msg) {
+    log.error(TAG_, 'An error ocurred on `amp-user-notification` ' +
+        `with html element id "${this.elementId_}". message: ${msg}`);
+    return false;
+  }
+
+  /**
+   * Get async cid service.
+   * @return {!Promise}
+   * @private
+   */
+  getAsyncCid_() {
+    return cidFor(this.win_).then(cid => {
+      // `amp-user-notification` is our cid scope, while we give it a resolved
+      // promise for the 2nd argument so that the 3rd argument (the
+      // persistentConsent) is the one used to resolve getting
+      // the external CID.
+      return cid.get('amp-user-notification',
+          Promise.resolve(), this.dialogPromise_);
+    });
+  }
+
+  /**
+   * @return {!Promise<boolean>}
+   * @private
+   **/
+  getIfHref_() {
+    if (!this.getIfHrefPromise_) {
+      this.getIfHrefPromise_ = this.getAsyncCid_()
+          .then(this.postShowEndpoint_.bind(this))
+          .then(this.onPostShowEndpointSuccess_.bind(this))
+          .catch(this.onRejection_.bind(this));
+    }
+    return this.getIfHrefPromise_;
+  }
+
+  /** @override */
+  shouldShow() {
+    return this.getIfHref_();
+  }
+
+  /**
+   * Shows the current user notification. Adds the `.amp-active` class to
+   * the element.
+   */
+  show() {
+    this.element.style.display = 'block';
+    this.element.classList.add('amp-active');
+  }
+
+  /**
+   * Hides the current user notification and invokes the `dialogResolve_`
+   * method. Removes the `.amp-active` class from the element.
+   */
+  dismiss() {
+    this.element.style.display = 'none';
+    this.element.classList.remove('amp-active');
+    this.element.classList.add('amp-hidden');
+    this.dialogResolve_(false);
+    this.postDismissEnpoint_();
+  }
+
+  /** @override */
+  isDismissed() {
+    return this.dialogPromise_;
+  }
+}
+
+
+/**
+ * UserNotificationService handles `amp-user-notification`
+ * queuing and registration, as well as exposing the components
+ * dismiss promise.
+ */
+export class UserNotificationService {
+
+  constructor(window) {
+    this.win_ = window;
+
+    /** @private @const {!Object<!UserNotificationDefer>} */
+    this.deferRegistry_ = Object.create(null);
+
+    /** @private @const {!Viewer} */
+    this.viewer_ = viewerFor(this.win_);
+
+    /** @private @const {?function(*)} */
+    this.documentReadyResolve_ = null;
+
+    /** @private @const {!Promise} */
+    this.documentReadyPromise_ = new Promise((resolve) => {
+      this.documentReadyResolve_ = resolve;
+    });
+
+    // TODO: Add `whenDocumentReady` function in document-state
+    // that returns a promise.
+    // See https://github.com/ampproject/amphtml/issues/1226
+    onDocumentReady(this.win_, () => {
+      this.documentReadyResolve_();
+    });
+
+    /** @private {!Promise} */
+    this.nextInQueue_ = this.viewer_.isEmbedded() ?
+        this.viewer_.whenVisible() : this.documentReadyPromise_;
```

Comment:
Once this service is "started" (DOMReady+Visible) - it could be a good time to check that all of the requested notifications are actually in DOM.